File: psx.R
## Pakistan Stock Exchange

library(quantmod)
library(tidyverse)

# Get the data

psx <- read_csv("data/psx.csv")

dim(psx)

glimpse(psx)

## Convert the date column to a date object where month day and year are separated by a hyphen

## psx$date <- as.Date(psx$Date, format = "%m-%d-%Y")

## Why above is incorrect

psx$date <- as.Date(psx$Date, format = "%m/%d/%Y")


glimpse(psx)

## Clean names using janitor

psx <-  janitor::clean_names(psx) |> select(-"date") |>
  rename(date = date_2)

## Create a new column called direction which will be the direction of the stock
## price movement when change_price is greater than 0 the direction is up

psx$direction <- ifelse(psx$change_price > 0, "Up", "Down")

## Convert change_price to a numeric

psx$change_price <- as.numeric(psx$change_price)

psx$change_percent <- as.numeric(gsub("%", "", psx$change_percent))

## Now creat direction column

psx$direction <- ifelse(psx$change_percent > 0, "Up", "Down")

## Arrange the data by date in ascending order

psx <- psx %>% arrange(date)

glimpse(psx)

## Create lag1 to lag5 columns using change_percent

psx <- psx %>% mutate(lag1 = lag(change_percent, 1),
                      lag2 = lag(change_percent, 2),
                      lag3 = lag(change_percent, 3),
                      lag4 = lag(change_percent, 4),
                      lag5 = lag(change_percent, 5))



# Convert `vol` to numeric
psx$vol <- as.numeric(gsub("M", "e6", gsub("K", "e3", psx$vol)))


## Logistic regression

glm.fit <- glm(direction ~ lag1 + lag2 + lag3 + lag4 + lag5 + vol,
                data = psx, family = binomial)

psx$direction <- ifelse(psx$direction == "Up", 1, 0)


psx_clean <- psx |> na.omit()

glm_fit <- glm(direction ~ lag1 + lag2 + lag3 + lag4 + lag5 + vol,
                data = psx_clean, family = binomial)

summary(glm_fit)


glm.probs=predict(glm_fit,type="response")
glm.probs[1:5]
glm.pred=ifelse(glm.probs>0.5,"Up","Down")

table(glm.pred,psx_clean$direction)
mean(glm.pred==psx_clean$direction)

# Make train and test data before 2024 and after 2024

train <- psx_clean$date < as.Date("2024-01-01")




glm_train <- glm(direction ~ lag1 + lag2 + lag3 + lag4 + lag5 + vol,
                data = psx_clean[train,], family = binomial)

## Predict on the test data

glm_probs <- predict(glm_train, newdata = psx_clean[!train,], type = "response")

glm_pred <- ifelse(glm_probs > 0.5, "Up", "Down")

table(glm_pred, psx_clean$direction[!train])

mean(glm_pred == psx_clean$direction[!train])

File: psx.py
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
import matplotlib.pyplot as plt

# Load the data
psx = pd.read_csv("data/psx.csv")

# Convert date column FIRST (before cleaning column names)
psx['date_clean'] = pd.to_datetime(psx['Date'], format='%m/%d/%Y')

# Clean column names
psx.columns = psx.columns.str.lower().str.replace(' ', '_')

# Drop the original 'date' column (which was 'Date' before cleaning)
# Keep 'date_clean' and rename it to 'date'
if 'date' in psx.columns:
    psx = psx.drop('date', axis=1)
psx = psx.rename(columns={'date_clean': 'date'})

# Clean the 'change_%' column (remove % sign and convert to numeric)
psx['change_%'] = psx['change_%'].str.replace('%', '').astype(float)

# Clean the volume column (remove 'K', 'M' suffixes and convert)
def clean_volume(vol_str):
    if isinstance(vol_str, str):
        vol_str = vol_str.replace('-', '0')  # Handle missing values
        if 'K' in vol_str:
            return float(vol_str.replace('K', '')) * 1000
        elif 'M' in vol_str:
            return float(vol_str.replace('M', '')) * 1000000
        else:
            return float(vol_str)
    return vol_str

psx['vol.'] = psx['vol.'].apply(clean_volume)

# Sort by date (now it exists!)
psx = psx.sort_values('date').reset_index(drop=True)

# Create direction variable (1 for Up, 0 for Down)
psx['direction'] = (psx['change_%'] > 0).astype(int)

# Create lagged variables for prediction
psx['lag1'] = psx['change_%'].shift(1)
psx['lag2'] = psx['change_%'].shift(2)
psx['lag3'] = psx['change_%'].shift(3)
psx['lag4'] = psx['change_%'].shift(4)
psx['lag5'] = psx['change_%'].shift(5)

# Remove rows with missing values (due to lagging)
psx_clean = psx.dropna()

print("Data shape after cleaning:", psx_clean.shape)
print("Direction distribution:\n", psx_clean['direction'].value_counts())

# Split data into training and testing (pre-2024 vs 2024)
psx_clean['year'] = psx_clean['date'].dt.year
train_data = psx_clean[psx_clean['year'] < 2024]
test_data = psx_clean[psx_clean['year'] == 2024]

print(f"\nTraining data shape: {train_data.shape}")
print(f"Test data shape: {test_data.shape}")

# Prepare features and target
feature_cols = ['lag1', 'lag2', 'lag3', 'lag4', 'lag5']
X_train = train_data[feature_cols]
y_train = train_data['direction']
X_test = test_data[feature_cols]
y_test = test_data['direction']

# Fit logistic regression model
log_model = LogisticRegression(max_iter=1000)
log_model.fit(X_train, y_train)

# Make predictions
y_pred_train = log_model.predict(X_train)
y_pred_test = log_model.predict(X_test)

# Calculate accuracies
train_accuracy = accuracy_score(y_train, y_pred_train)
test_accuracy = accuracy_score(y_test, y_pred_test)

print(f"\nTraining Accuracy: {train_accuracy:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")

# Confusion Matrix for test data
cm = confusion_matrix(y_test, y_pred_test)
print("\nConfusion Matrix (Test Data):")
print(cm)

# Classification Report
print("\nClassification Report (Test Data):")
print(classification_report(y_test, y_pred_test, target_names=['Down', 'Up']))

# Display model coefficients
print("\nModel Coefficients:")
for i, coef in enumerate(log_model.coef_[0]):
    print(f"Lag{i+1}: {coef:.4f}")
print(f"Intercept: {log_model.intercept_[0]:.4f}")

# Plot predictions vs actual
plt.figure(figsize=(12, 6))
plt.plot(test_data['date'], y_test.values, label='Actual Direction', alpha=0.7, marker='o')
plt.plot(test_data['date'], y_pred_test, label='Predicted Direction', alpha=0.7, marker='x')
plt.xlabel('Date')
plt.ylabel('Direction (0=Down, 1=Up)')
plt.title('PSX Direction: Actual vs Predicted (2024)')
plt.legend()
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

# Display first few rows
print("\nFirst few rows of cleaned data:")
print(psx_clean.head())
